Support for Python  version 3.12 and 3.13

DivingDuck opened the pull request:
Build with 3.13 is still experimental. It builds the binaries but due to compiling errors in library pyglet version 1.5.29 we miss the support for 3D visualization. Beside this Pronterface and Pronsole is fully functional.

**Changes in release_windows.bat**
1. Change pyglet==1.5.27 to pyglet==1.5.29
2. Add Python 3.12 and 3.13 as build option, remove support for Python 3.7
   Python 3.13 is still experimental as there is for now no 3D visualization available because of incompatibility issues with pyglet
3. Add library pytest to additional libraries
4. Add removal of gcoder_line.cp???-win_amd??.pyd files with python versions that have 3 instead of 2 digits
5. Remove folder \build in addition as it will not removed any longer automatically in the build process
6. Remove manual fix for pillow==9.5 as no longer needed

**CleanCacheFiles.bat**
This batch file will clean up all __pycache__ folders, *.pyd files and left over gcoder_line files from previous builds. For local Windows only. It is helpful when you switch between Python versions or you have problems with incompatible library files or versions.

**Changes in requirements.txt**
Remove dedicated version for pillow in Windows build environment
`pillow < 10.0; sys_platform == 'win32'`
--> to: `pillow ;sys_platform == 'win32'`
 
**Changes in buildpackage-win.yml**
Add support for Python 3.12 and 3.13 and mark Python 3.13 as experimental

**Various fixes for visible SyntaxError warnings of invalid escape sequence when compiling with Python 3.12 and 3.13**
(in question to become in Python >3.13 a syntax error)

See https://docs.python.org/dev/whatsnew/3.6.html#deprecated-python-behavior:
_A backslash-character pair that is not a valid escape sequence now generates a DeprecationWarning. 
Although this will eventually become a SyntaxError, that will not be for several Python releases. 
(Contributed by Emanuel Barry in bpo-27364.)_

and https://docs.python.org/3/reference/lexical_analysis.html, Chapter 2.4.1.1. Escape sequences:
_Changed in version 3.6: Unrecognized escape sequences produce a DeprecationWarning.
Changed in version 3.12: Unrecognized escape sequences produce a SyntaxWarning. In a future Python version they will be eventually a SyntaxError._
**Solution:** Change string to raw string when escape sequences are involved

**\printrun\pronterface.py:2099:** SyntaxWarning: invalid escape sequence '\d'
`self.sdfiles.append(re.sub(" \d+$", "", line.strip().lower()))  # NOQA`
--> to: `self.sdfiles.append(re.sub(r" \d+$", "", line.strip().lower()))  # NOQA`

**\printrun\pronsole.py:61**: SyntaxWarning: invalid escape sequence '\d'
`tempreading_exp = re.compile('\\bT\d*:')`
--> to: `tempreading_exp = re.compile(r'\\bT\d*:')`

**\printrun\pronsole.py:201:** SyntaxWarning: invalid escape sequence '\d'
`self.lineignorepattern=re.compile("ok ?\d*$|.*busy: ?processing|.*busy: ?heating|.*Active Extruder: ?\d*$")`
--> to: `self.lineignorepattern = re.compile(r"ok ?\d*$|.*busy: ?processing|.*busy: ?heating|.*Active Extruder: ?\d*$")`

**\printrun\pronsole.py:1140:** SyntaxWarning: invalid escape sequence '\d'
`self.sdfiles.append(re.sub(" \d+$","",line.strip().lower()))`
--> to: `self.sdfiles.append(re.sub(r" \d+$","",line.strip().lower()))`

**\printrun\gcoder.py:28:** SyntaxWarning: invalid escape sequence '\('
`gcode_exp = re.compile("\([^\(\)]*\)|;.*|[/\*].*\n|([%s])\s*([-+]?[0-9]*\.?[0-9]*)" % to_parse)`
--> to: `gcode_exp = re.compile(r"\([^\(\)]*\)|;.*|[/\*].*\n|([%s])\s*([-+]?[0-9]*\.?[0-9]*)" % to_parse)`

**\printrun\gcoder.py:29:** SyntaxWarning: invalid escape sequence '\('
`gcode_strip_comment_exp = re.compile("\([^\(\)]*\)|;.*|[/\*].*\n")`
--> to: `gcode_strip_comment_exp = re.compile(r"\([^\(\)]*\)|;.*|[/\*].*\n")`

**\printrun\gcoder.py:30:** SyntaxWarning: invalid escape sequence '\('
`m114_exp = re.compile("\([^\(\)]*\)|[/\*].*\n|([XYZ]):?([-+]?[0-9]*\.?[0-9]*)")`
--> to: `m114_exp = re.compile(r"\([^\(\)]*\)|[/\*].*\n|([XYZ]):?([-+]?[0-9]*\.?[0-9]*)")`

**\printrun\gcoder.py:31:** SyntaxWarning: invalid escape sequence '\('
`specific_exp = "(?:\([^\(\)]*\))|(?:;.*)|(?:[/\*].*\n)|(%s[-+]?[0-9]*\.?[0-9]*)"`
--> to:  `specific_exp = "(?:\([^\(\)]*\))|(?:;.*)|(?:[/\*].*\n)|(%s[-+]?[0-9]*\.?[0-9]*)"`

**\printrun\device.py:209:** SyntaxWarning: invalid escape sequence '\.'
`host_regexp = re.compile("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$|^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$")`   
--> to: `host_regexp = re.compile("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$|^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$")`

**\printrun\gcodeplater.py:48:** SyntaxWarning: invalid escape sequence '\.'
`rewrite_exp = re.compile("(%s)" % "|".join(["X([-+]?[0-9]*\.?[0-9]*)",`
--> to: `rewrite_exp = re.compile("(%s)" % "|".join([r"X([-+]?[0-9]*\.?[0-9]*)",`

**\printrun\gcodeplater.py:49:** SyntaxWarning: invalid escape sequence '\.'
`"Y([-+]?[0-9]*\.?[0-9]*)"]))`
--> to: `r"Y([-+]?[0-9]*\.?[0-9]*)"]))`

**Can't be fixed because external library and do not harm:**
**\v3\Lib\site-packages\pyglet\gl\wgl.py:36:** SyntaxWarning: invalid escape sequence '\c'
`"""Wrapper for C:\cygwin\home\Alex\pyglet\tools\wgl.h`

DivingDuck commented:
For now this PR is work in progress.

rockstorm101 commented:
> For now this PR is work in progress.

Thank you for such a detailed PR. Let me know when it's ready for review/merge.

DivingDuck commented:
Hi @rockstorm101,

Some good news. It looks like we can run Pronterface with Python 3.13 w/o the experimental free-threaded mode as well. 🙂

Plyglet 1.5 became an update some days ago to version 1.5.30 (https://github.com/pyglet/pyglet/pull/1243) which make the 3D visualization working for Python 3.13 again.
I did some tests for all supported python versions from 3.8 up to 3.13 in x32 and x64 with the new version of pyglet to be sure we don't come in trouble.

**Test with Windows 10:**
|Py version| x86 | x64 |
|----------|-----|-----|
|3.13.1t| no | no |
|3.13.1| ok | ok |
|3.12.8| ok | ok |
|3.11.9| ok | ok |
|3.10.11| ok | ok |
|3.9.13| ok | ok |
|3.8.10| no | ok |

**Latest Windows Artifacts:** https://github.com/kliment/Printrun/actions/runs/12306812910

**Additional changes for Windows:**
release_windows.bat and windows workflow: Remove remark for experimental support for Python 3.13 and add that the experimental Python 3.13t is not supported for now.
release_windows.bat: remove library pypiwin32 and add pywin32 as this version seems to be more supported and compatible with all our supported Python versions for windows. https://pypi.org/project/pywin32/ , https://pypi.org/project/pypiwin32/

**MacOs:**
macos-12 is now no longer supported. See issue #1461. I removed macos-12 from the macOS workflow and add macos-13, macos-14 and macos-15 for testing.
In addition I remove Python 3.10 as this version isn't compatible with macos-14 and newer. Therefore I add Python 3.11, 3.12 and 3.13 for testing.
Observation from the logs: The workflow step of Build Cython ext contains some warnings. 
I can't test those versions as I haven't a Mac available.
**Latest macOS Artifacts:** https://github.com/kliment/Printrun/actions/runs/12306812915

**Build-wheels:**
I changed this workflow as well for macOS. Removed macos-12 and add macos-13
**Latest build-wheels Artifacts:** https://github.com/kliment/Printrun/actions/runs/12306812905
Test needed for Linux as well. Is there something to do in addition for the Linux part?

After testing we need to decide what versions we want to remain for all workflows.

neofelis2X commented:
Hello @DivingDuck,
thank you for updating the builds. I tested some of the artifacts on a macOS 15.1.1 and on a macOS 12.7.6

- Surprisingly, the macos-13 artifacts run on my old macos12 macbook. The later builds **do not**.  That is expected.

The modern macOS is a different story:
- All artifacts _work_ on macos 15. But macos _really does not want to open_ them.
- First I get this message and with no option to open them at all:
<img width="262" alt="Screenshot 2024-12-13 at 14 18 44" src="https://github.com/user-attachments/assets/9c0ff394-0678-4427-816c-6131c554396a" />

- Then I had to dig in the settings to find this option. There was no hint anywhere that I have this option as a user.
<img width="582" alt="Screenshot 2024-12-13 at 14 13 13" src="https://github.com/user-attachments/assets/9e8eb6e1-3ab0-4700-b9ec-4d829f160fc7" />

- After clicking on 'Open Anyway' it just works.

That means we will have to improve the instructions on how to install and open Pronterface on mac.

DivingDuck commented:
@neofelis2X,
thank you very much taking the time for testing. Indeed, we need to improve the documentation for macOS then.
Guess we should build in addition a matrix of what platform, macOS- and python version works and what not. Something like this if it makes sense:

| Platform | Python | x64 | macOS 12.7.6 <sup>1</sup> | macOS 15.1.1 |
|----------|--------|-----|---------------------------|----------------|
|macos-13|py 3.10| ok | ok | no |
|macos-13|py 3.11| ok | ok | no |
|macos-13|py 3.12| ok | ok | no |
|macos-13|py 3.13| ok | ok | no |
|macos-14|py 3.10| no | no |  ?  |
|macos-14|py 3.11| ok | no |  ?  |
|macos-14|py 3.12| ok | no |  ?  |
|macos-14|py 3.13| ok | no |  ?  |
|macos-15|py 3.10| no | no | no |
|macos-15|py 3.11| ok | no | ok |
|macos-15|py 3.12| ok | no | ok |
|macos-15|py 3.13| ok | no | ok |

<sup>1</sup> Tested with Macbook

neofelis2X commented:
Platform | Python | x64 | macOS 12.7.6 [^1]  | macOS 15.1.1[^2]
-- | -- | -- | :--: | :--:
macos-13 | py 3.10 | ok | untested | untested
macos-13 | py 3.11 | ok | yes | yes
macos-13 | py 3.12 | ok | yes | yes
macos-13 | py 3.13 | ok | yes | yes
macos-14 | py 3.10 | no | no | no
macos-14 | py 3.11 | ok | no | yes
macos-14 | py 3.12 | ok | no | yes
macos-14 | py 3.13 | ok | no | yes
macos-15 | py 3.10 | no | no | no
macos-15 | py 3.11 | ok | no | yes
macos-15 | py 3.12 | ok | no | yes
macos-15 | py 3.13 | ok | no | yes

[^1]: Tested on a 2015 _Intel_ Macbook
[^2]: Tested on a 2022 _M2_ Macbook

Okay, so that would be the table from my side. Which versions / combinations do you want to release?

I also started to write a little manual on how to handle the security warnings.

DivingDuck commented:
Hi @neofelis2X,

Sorry for my late answer. A flue catch me last week.
Regarding combinations I think we should provide two versions to have a brighter basis for macOS users. MacOS-13 with actual Python 3.13 and macOS-15 with Python 3.13.

Python 3.13 because of Pythons fast development cycle and it seems to be fine for all macOS versions so far.

What do you think about this choice? 

Great that you are already start writing a little manual for the installation. +1

I was thinking about Python for the Windows in the same direction - one version with actual version and maybe one with an older version. All other combinations can be build locally from a user as I provide this in the local build script as already shown [above](https://github.com/kliment/Printrun/pull/1465#issuecomment-2541629148).


@rockstorm101, 
Any suggestions from your side, especially for Linux platforms?

rockstorm101 commented:
> Regarding combinations I think we should provide two versions to have a brighter basis for macOS users. MacOS-13 with actual Python 3.13 and macOS-15 with Python 3.13.

For pre-built binaries that are distributed to users I would aim for the lowest version of OS plus the lowest version of Python possible. Because that binary will work (correct me if I'm wrong here) on every OS version above that. Essentially I'm trying to make things simple for users. If one binary works for everyone, just give them one binary and users won't have to decide which binary they have to download. If not, let's try to reduce the number of binaries to the very minimum.


> Any suggestions from your side, especially for Linux platforms?

The current recommended way of installing on Linux is via PyPI and there we upload every possible combination so I don't see a debate there. If we fix the pyglet issue, then linux distro packages will create their packages for whatever versions work for them, again, no need for us (upstream) to worry about it I think. So, all good from Linux point of view.

neofelis2X commented:
No problem, I was also quite occupied these days. I wish you a speedy recovery then!

>  I would aim for the lowest version of OS plus the lowest version of Python possible. [...] Essentially I'm trying to make things simple for users.

I generally agree with that, one binary would probably work well for everyone. 
However, I did some tests with the versions and compared `macos13-python3.11` with `macos14-python3.13` : The newer version has a 40% smaller filesize, starts almost 50% faster and also seems to open files about 20% - 30% faster. That are some pretty nice benefits.
On that basis I propose to combine both of your suggestions and release `macos13-python3.11` for older macs and `macos14-python3.13` for modern macs.


> Great that you are already start writing a little manual for the installation. +1

How should I upload it? A separate PR that changes the README? Or as a new file, that we can also  include into the released .zip file?

neofelis2X commented:
Hi, sorry for the long silence. I got a bit distracted.

> Or does it come by default with that Python version?

Python is packaged within the macos application. So it doesnt need to be installed. :)

> GitHub's Wiki pages which could also be an option.

Yes, I checked out the wiki and found it to be a good place for this instructions. But now I can not upload my changes because I have no access rights. And wikis don't support regular PRs, it seems?

 Can I get rights to upload to the wiki?

DivingDuck commented:
Hi,
Sorry for my late answer. I enjoyed a 3 weeks break w/o looking to any kind of emails - what a pleasure. 
Now I'm able to go (slowly) to all the open ends and emails.

Reminder to myself:
Don't forget to update translation files (https://github.com/kliment/Printrun/pull/1473#issuecomment-2543065514) before this PR will be implemented

DivingDuck commented:
Regarding Python, I want to settle this

> Sounds good to me. Maybe not all macOS 14 users have Python 3.13? Or does it come by default with that Python version?

I don't know. On the other side Py3.13 is the actual standard version since Oct. 2024 and as mentioned it have some benefits for the actual macOS.

I have the same situation for Window and Visual Studio 2022. Python 3.13 is not official supported but compile correct and Visual Studio Code do support that version. However when I look to Python 3.8, Visual studio compile the x64 but not x86 locally and with Github I can compile both version flawlessly with the build in VS 2022.

Anyway, here is my suggestion:

**MacOS:**
`macos13-python3.11` for older macs and `macos14-python3.13`

**Windows:**
`Python 3.8` as this is now really the last time for `x86/x64` (as last chance for really old Windows machines) and `Python 3.13` for Windows 10 and 11
Remark: I want to get rid of all EOL Python versions for Windows as soon as possible. Ideally we should support 3.12 as min Version for the future. Who want to use older versions should build and test them locally. 

**Linux:**
What ever is supported via pypi

rockstorm101 commented:
> Can I get rights to upload to the wiki?

You should have them know. Thanks for contributing! :)

> here is my suggestion:
> 
> **MacOS:** `macos13-python3.11` for older macs and `macos14-python3.13`
> 
> **Windows:** `Python 3.8` as this is now really the last time for `x86/x64` (as last chance for really old Windows machines) and `Python 3.13` for Windows 10 and 11 Remark: I want to get rid of all EOL Python versions for Windows as soon as possible. [...]
> 
> **Linux:** What ever is supported via pypi

Sounds good to me. Let's go with that.

> Ideally we should support 3.12 as min Version for the future. Who want to use older versions should build and test them locally.

We can definitely ditch Python 3.8 since it is no longer supported anyway. However I would do as you suggested and keep support for the lowest Python version possible to help people with old hardware. Nonetheless, if packaging effort becomes a burden, we must of course optimize our little spare time and maybe just support the latest stable Python. So up to you really.

neofelis2X commented:
> You should have them know. Thanks for contributing! :)

Thank you very much! Wiki is updated and #1479 is pushed.

DivingDuck commented:
Latest updates:
- Change default Python version to 3.13 for Windows local builds
- update build workflow for Windows binaries, restrict builds to Python 3.8 and 3.13
- update build workflow for macOS binaries, restrict builds to macos13/14 with python3.11/3.13
- update translation files
- update readme.md, change Python 3.10 to 3.13 as default for Windows builds and add some more information for the Windows development environment.

Missing:
Update translation files when #1480 is implemented

DivingDuck commented:
Hi @rockstorm101 and @neofelis2X,

I did some more updates:
- update build workflow for macOS binaries:
  - restrict platform builds [Restrict plattform builds for macos-13 to Python 3.11 and macos-14 to Python 3.13 only
  - add translation files to macOS distribution package (as we have for Windows
- readme.md: Add new chapter Translating Pronterface

It would be nice if you can take a look on the new chapter [()](https://github.com/DivingDuck/Printrun/blob/PythonSupport3.12_3.13/README.md) and give me a feedback for any changes needed 🙂

This should then close issue #1429 as well.

There is another point that came in my mind again. @rockstorm101 and I had a conversation quite some time ago where we thought about adding the file readme.md to the binary distribution packages. Do you think we should still do this?

rockstorm101 commented:
> about adding the file readme.md to the binary distribution packages. Do you think we should still do this?

Short answer: if it is simple to add, I would add it.

Long answer: In the old days, the norm was that binary distributed packages should be "self-contained". Meaning that a user that got such package should not require internet or any other resource to make use of the software. Which means the documentation had to be packaged with the software so the user could understand and use it. In the modern days where everyone's got a phone with internet access I believe such norm could be relaxed really. But... old habits die hard ;)

rockstorm101 commented on README.md:
```diff
@@ -223,6 +226,64 @@ You need to adjust the path to your location of pronsole.exe.
+
+Pronterface supports the localization of the GUI in your language. We would be more
+than happy if users support us with updated translation files in all different
+languages. Actual we have translation files for AR, DE, FR, HY, IT and NL available,
```
Actual -> Actually

DivingDuck replied:
done

rockstorm101 commented on README.md:
```diff
@@ -223,6 +226,64 @@ You need to adjust the path to your location of pronsole.exe.
+Poedit (https://poedit.net/). The software is available for Linux, macOS and Windows.
+
+The translation files are already included in the distributed pre-compiled binaries
+for Windows. You need only to open the pronterface.po of the corresponding language
```
"for Windows and macOS"? Once this PR is merged

DivingDuck replied:
done

rockstorm101 commented on README.md:
```diff
@@ -223,6 +226,64 @@ You need to adjust the path to your location of pronsole.exe.
 
 Run 'release_windows.bat' when ever you make changes or updates. With each new run it will compile the binaries and update all involved libraries in the virtual environment if needed. Delete the virtual environment if you have problems with it.
 
+### Translating Pronterface
```
I feel like this should be a level 2 header instead.

DivingDuck replied:
Indeed, good point. Done

rockstorm101 commented on README.md:
```diff
@@ -223,6 +226,64 @@ You need to adjust the path to your location of pronsole.exe.
+#### How to update a translation
+
+The best way is to clone first our repository and then work on the translation file
+local. Send us a pull request when you have finish your work. Doing the translation
```
finish -> finished

DivingDuck replied:
done.

rockstorm101 commented on README.md:
```diff
@@ -223,6 +226,64 @@ You need to adjust the path to your location of pronsole.exe.
+The best way is to clone first our repository and then work on the translation file
+local. Send us a pull request when you have finish your work. Doing the translation
+in steps is also OK, please mention it in the PR. We will then wait with releasing
+your update up to the point you tell us you have finish your work.
```
finish -> finished

DivingDuck replied:
done

rockstorm101 commented on README.md:
```diff
@@ -223,6 +226,64 @@ You need to adjust the path to your location of pronsole.exe.
+
+The translation is organized in the folder `\locale` and includes sub folders for each
+supported language. The structure is mandatory and should not be changed. The file
+`pronterface.po` contains all translatable strings and is the file to be edit for a
```
edit -> edited

DivingDuck replied:
done

rockstorm101 reviewed (commented):
Thanks for enhancing the documentation!

rockstorm101 commented on printrun/pronsole.py:
```diff
@@ -402,9 +402,9 @@ def default(self, l):
         else:
             if self.processing_rc:
-                logging.debug(f"Ignored configuration command '{l}'")
+                logging.debug(_("Ignored configuration command") + f" '{l}'")
```
I wasn't aware of the problems regarding translating f-strings. I'm sorry I've used them all over the place. Should we fall back to the str.format() style? E.g. make this section look like the following. Feels more elegant than splitting strings but I'm up for discussion.
```python
        else:
            if self.processing_rc:
                logging.debug(_("Ignored configuration command '{}'"
                                .format(l)))
            else:
                self.log(_("Unknown command '{}'".format(l)))
```

DivingDuck replied:
Yes, this is common everywhere these days and I was as well surprised and only recognize this when I tried building the new pot file. Nothing to be sorry about 🙂

I will change as you mention and update translation files

neofelis2X commented:
The artifacts work well and I dont have additional comments on the changes. 👌

But I am confused about the translation. I tried to get  Printrun to run in German, compiled the .po into a .mo with Poedit and tried different locations for the locals folder. But no success, it always stays English.

Are the translations expected to work already?
And what would be my steps to use them?

DivingDuck commented:
Hi @neofelis2X,
on Windows I copy the folder locale to the folder where the binaries of Pronterface are stored and Pronterface then find the needed .mo file based on the system language of your system in the locale folder.

```
.\Pronterface
   .\locale
     .\de
       .\LC_MESSAGES
         pronterface.po
         pronterface.mo
     .\it
       .\LC_MESSAGES
         pronterface.po
         pronterface.mo
     ...
    pronterface.pot
   Pronterface.exe
   Pronsole.exe
   Plater.exe
```
Three things worth to check:

- What happen if you run Pronterface from source with a copy of the .mo file in the locale folder?
- Copy the `.mo` file in the directory that represent your language in case your system language is different to the language file you want to test ( e.g. your system language is English: create a folder structure for .\en and copy the folder `.\LC_MESSAGES`  from `.\de into .en\`
- If I remember correct Linux and macOS save translations on a different place, was it a shared folder?

**Edit after taking a look in the code**: 
Take a look on` .\printrun\utils.py` line 41 ff and there at line 47:

```
if osPlatform == "Darwin":
    # improvised workaround for macOS crash with gettext.translation, see issue #1154
    if os.path.exists(shared_locale_dir):
        gettext.install(domain, shared_locale_dir)
    else:
        gettext.install(domain, './locale')
else:
    if os.path.exists('./locale'):
        translation = gettext.translation(domain, './locale',
                                          languages=[lang[0]], fallback= True)
    else:
        translation = gettext.translation(domain, shared_locale_dir,
                                          languages=[lang[0]], fallback= True)
    translation.install()
```
So maybe something like `usr/share/locale` ?  Link to issue #1154

neofelis2X commented:
Thank you for the hints, that helped a lot to pin down the problem. The translations do work, but printrun is looking for them at the wrong location, in this case.

macOS has a folder `/usr/share/locale/` but for regular users is it read-only and I can't just copy files there.

Usually the localisation files are in the _Resources_ folder of the Application-bundle.

<details>

<summary>Open to see example</summary>

The Resources folder of the app _Handbrake_.

![res_hb](https://github.com/user-attachments/assets/f8f6d56f-93af-469e-acae-f0bdfe99c073)

</details>

As a quick fix this lines:
``` python
if osPlatform == "Darwin":
    # improvised workaround for macOS crash with gettext.translation, see issue #1154
    if os.path.exists(shared_locale_dir):
        gettext.install(domain, shared_locale_dir)
    else:
        gettext.install(domain, './locale')
```

could be changed like this:
``` python
if osPlatform == "Darwin":
    # improvised workaround for macOS crash with gettext.translation, see issue #1154
    gettext.install(domain, './locale')
```

Then it would work just the same as it does for windows now.

DivingDuck commented:
> could be changed like this:
> 
> ```python
> if osPlatform == "Darwin":
>     # improvised workaround for macOS crash with gettext.translation, see issue #1154
>     gettext.install(domain, './locale')
> ```
> 
> Then it would work just the same as it does for windows now.

Thanks a lot for testing. I like your solution and will change the code. This makes the handling with translations more simple.

DivingDuck commented:
**New updates**:

- Merge latest changes from master branch
- pronsole.py:
     Update the log string as mentioned
- utils.py: 
     macOS: search translation folder and files on the same place where the binary file is located, it acts now same as for Windows
- README.md:
     Update chapter Translating Pronterface, add information where Pronterface look for translation files
- Update all translation files:
    Because of code changes and change one menu string for platter in German translation to be more specific.
- release_windows.bat:
    Add copy README.md to the distribution folder

**Latest artifacts**:
Windows: https://github.com/kliment/Printrun/actions/runs/13393302186
MacOS: https://github.com/kliment/Printrun/actions/runs/13393302187
Wheels: https://github.com/kliment/Printrun/actions/runs/13393302161

For me the PR is now ready to merge if there are no further findings. Please check the latest changes. Thank you very much for your help @rockstorm101 & @neofelis2X

rockstorm101 commented:
Thanks for your efforts here. Unless @neofelis2X wants to add anything I'm happy to merge now.

neofelis2X commented:
Nope, i have nothing to add, looks great!

rockstorm101 commented:
Thanks for your work on this one @DivingDuck. I would simply suggest to maybe split the work into several simpler/shorter PRs for the future, one for each topic (e.g. syntax fix, CI stuff, translations,...). I feel like it keeps things tidier and more modular and does not tie one thing to the other. It also makes reviews simpler. If one big PR works better for you though, keep it like that. This is just a suggestion.

DivingDuck commented:
@DivingDuck, I try to do so next time. 😇
I thought the same somewhen beginning this month, but was also surprised about all the changes needed for the change to py 3.13.
